graphrag/storage/file_pipeline_storage.py

```diff
@@ -74,21 +88,123 @@ def item_filter(item: dict[str, Any]) -> bool:
             if progress is not None:
                 progress(_create_progress_status(num_loaded, num_filtered, num_total))
 
+    def find_s3(
+        self,
+        file_pattern: re.Pattern[str],
+        base_dir: str | None = None,
+        progress: Any
+        | None = None,  # Replace `ProgressLogger` with your custom logger if needed
+        file_filter: dict[str, Any] | None = None,
+        max_count=-1,
+    ) -> Iterator[tuple[str, dict[str, Any]]]:
+        """Find files in S3 storage using a file pattern and optional filters."""
+
+        def item_filter(item: dict[str, Any]) -> bool:
+            if file_filter is None:
+                return True
+            return all(re.match(value, item[key]) for key, value in file_filter.items())
+
+        search_prefix = f"{self._root_dir}/" if self._root_dir else ""
+
+        log.info(
+            "Searching S3 bucket '%s' with prefix '%s' for files matching pattern '%s'",
+            self.bucket_name,
+            search_prefix,
+            file_pattern.pattern,
+        )
+
+        num_loaded = 0
+        num_filtered = 0
+
+        try:
+            paginator = self.s3_client.get_paginator("list_objects_v2")
+            page_iterator = paginator.paginate(
+                Bucket=self.bucket_name, Prefix=search_prefix
+            )
+
+            for page in page_iterator:
+                if "Contents" not in page:
+                    continue
+
+                for obj in page["Contents"]:
+                    key = obj["Key"]
+                    match = file_pattern.match(key)
+                    if match:
+                        group = match.groupdict()
+                        if item_filter(group):
+                            filename = key[len(search_prefix) :]
+                            yield (filename, group)
+                            num_loaded += 1
+                            if max_count > 0 and num_loaded >= max_count:
+                                return
+                        else:
+                            num_filtered += 1
+                    else:
+                        num_filtered += 1
+
+                    if progress is not None:
+                        progress({
+                            "num_loaded": num_loaded,
+                            "num_filtered": num_filtered,
+                            "current_file": key,
+                        })
+
+        except ClientError:
+            log.exception("Failed to list objects in S3")
+            raise
+
+
     async def get(
         self, key: str, as_bytes: bool | None = False, encoding: str | None = None
     ) -> Any:
         """Get method definition."""
         file_path = join_path(self._root_dir, key)
 
         if await self.has(key):
```

Comment:
```suggestion
        if await self.has_s3(key):
```
是漏掉了嗎?
覆寫 `has()`、`get()`、`set()` 或許應該是更好的寫法

Reply:
不只這裡，其他地方也有相這樣使用 local file 的邏輯